### PR TITLE
Resize user images as celery task

### LINKED ISSUE #5664: Make resizing of user images a celery task
it may time out as it is being handled synchronously right now.

### PR DESCRIPTION
**Pull request by Haider8**
Fixes #5664 

#### Checklist

- [ ] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [ ] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Now  the user images can be resized asynchronously as a celery task.

### CODE REVIEW

**Comment by Haider8** [comment]
@iamareebjamal please review

**Comment by Haider8** [comment]
@iamareebjamal please review. Have made required changes.

**Comment by iamareebjamal** [comment]
@Haider8 What are the fields present in User model for avatar and its sizes?

**Comment by Haider8** [comment]
avatar_url, original_image_url, thumbnail_image_url, small_image_url, icon_image_url
@iamareebjamal

**Comment by Haider8** [comment]
@iamareebjamal From users schema 

```
avatar_url = fields.Url(allow_none=True)
original_image_url = fields.Url(allow_none=True)
thumbnail_image_url = fields.Url(dump_only=True, allow_none=True)
small_image_url = fields.Url(dump_only=True, allow_none=True)
icon_image_url = fields.Url(dump_only=True, allow_none=True)
```

**Comment by iamareebjamal** [comment]
@CosmicCoder96 Any clue how avatar_url and original_image_url differ and are used?

I think we should get rid of original_image_url and only allow avatar_url as it creates confusion

**Comment by iamareebjamal** [comment]
@Haider8 For now, only resize avatar_url and fill thumbnail_image_url, small_image_url, icon_image_url and ignore original_image_url

**Comment by Haider8** [comment]
@iamareebjamal please review

File: app/api/users.py [hunk 1]
```diff
@@ -87,8 +87,8 @@ def after_create_object(self, user, data, view_kwargs):
         #     del uploaded_images['large_image_url']
         #     self.session.query(User).filter_by(id=user.id).update(uploaded_images)
 
-        if data.get('original_image_url'):
-            start_image_resizing_tasks(user, data['original_image_url'])
+        if data.get('avatar_url'):
```
Comment (iamareebjamal) [inline app/api/users.py]: Do this after update as well if avatar URL has changed

**Comment by Haider8** [comment]
@CosmicCoder96 Yes, i have tried to upload image locally and it was working fine.

**Comment by CosmicCoder96** [comment]
@Haider8 Not the uploading part, the resizing.

**Comment by Haider8** [comment]
> @Haider8 Not the uploading part, the resizing.

Yes, i have also tried to resize the images. Working fine locally

**Comment by CosmicCoder96** [comment]
can you share the JSON response of an event whose images got resized? As the path addresses might be off.

File: app/api/helpers/tasks.py [hunk 2]
```diff
@@ -97,6 +97,22 @@ def resize_event_images_task(self, event_id, original_image_url):
         logging.exception('Error encountered while generating resized images for event with id: {}'.format(event_id))
 
 
+@celery.task(base=RequestContextTask, name='resize.user.images', bind=True)
+def resize_user_avatar_images_task(self, user_id, avatar_image_url):
+    user = safe_query(db, User, 'id', user_id, 'user_id')
+    try:
+        logging.info('User avatar image resizing tasks started {}'.format(avatar_image_url))
+        uploaded_images = create_save_image_sizes(avatar_image_url, 'user-image', user.id)
+        user.avatar_url = uploaded_images['avatar_url']
```
Comment (iamareebjamal) [inline app/api/helpers/tasks.py]: This should be set to `original_image_url`

My oversite on that part

Comment (Haider8) [reply app/api/helpers/tasks.py]: ```user.avatar_url``` ?

**Comment by Haider8** [comment]
> @Haider8 can you share the JSON response of an user whose images got resized? As the path addresses might be off.

@CosmicCoder96 
```
{
  "url": "http://127.0.0.1:5000/static/media/temp/images/fc8a3cd4-fc9d-47f8-a61a-058eabfe5d4f/Q2gyUjQxeW/004e1ae4-4f46-429b-943f-0c6939b67063.png"
}
```

**Comment by iamareebjamal** [comment]
@Haider8 This is not the JSON of a user.

**Comment by Haider8** [comment]
@CosmicCoder96 This is the JSON of the user whose image is resized
```
{
    "data": {
        "type": "user",
        "attributes": {
            "last-accessed-at": "2019-03-24T15:07:55.544025+00:00",
            "email": "haider.lee23@gmail.com",
            "is-user-organizer": true,
            "thumbnail-image-url": null,
            "is-user-attendee": false,
            "is-super-admin": true,
            "facebook-id": null,
            "is-user-coorganizer": false,
            "first-name": "Haider ",
            "original-image-url": null,
            "is-admin": true,
            "facebook-url": null,
            "was-registered-with-order": false,
            "is-marketer": false,
            "is-sales-admin": true,
            "contact": "7678463676",
            "details": "User details",
            "small-image-url": null,
            "is-user-registrar": false,
            "twitter-url": null,
            "last-name": "Ali",
            "created-at": "2019-01-17T17:58:59.432218+00:00",
            "icon-image-url": null,
            "is-user-moderator": false,
            "is-user-track-organizer": false,
            "is-verified": true,
            "instagram-url": null,
            "deleted-at": null,
            "google-plus-url": null,
            "avatar-url": "http://127.0.0.1:5000/static/media/temp/images/dfb8d0be-c3f3-41a1-a631-2c0b0a5b37de/ak05cGVaQl/3810d41c-4707-4166-a99b-f67fb57440f0.png"
        },
        "relationships": {
            "email-notifications": {
                "links": {
                    "self": "/v1/users/1/relationships/email-notifications",
                    "related": "/v1/users/1/email-notifications"
                }
            },
            "registrar-events": {
                "links": {
                    "self": "/v1/users/1/relationships/registrar-events",
                    "related": "/v1/events"
                }
            },
            "sessions": {
                "links": {
                    "self": "/v1/users/1/relationships/sessions",
                    "related": "/v1/users/1/sessions"
                }
            },
            "track-organizer-events": {
                "links": {
                    "self": "/v1/users/1/relationships/track-organizer-events",
                    "related": "/v1/events"
                }
            },
            "events": {
                "links": {
                    "self": "/v1/users/1/relationships/events",
                    "related": "/v1/users/1/events"
                }
            },
            "discount-codes": {
                "links": {
                    "self": "/v1/users/1/relationships/discount-codes",
                    "related": "/v1/users/1/discount-codes"
                }
            },
            "access-codes": {
                "links": {
                    "self": "/v1/users/1/relationships/access-codes",
                    "related": "/v1/users/1/access-codes"
                }
            },
            "alternate-emails": {
                "links": {
                    "self": "/v1/users/1/relationships/alternate-emails",
                    "related": "/v1/users/1/alternate-emails"
                }
            },
            "attendees": {
                "links": {
                    "self": "/v1/users/1/relationships/attendees",
                    "related": "/v1/users/1/attendees"
                }
            },
            "favourite-events": {
                "links": {
                    "self": "/v1/users/1/relationships/favourite-events",
                    "related": "/v1/users/1/user-favourite-events"
                }
            },
            "sales-admin-events": {
                "links": {
                    "self": "/v1/users/1/relationships/sales-admin-events",
                    "related": "/v1/events"
                }
            },
            "speakers": {
                "links": {
                    "self": "/v1/users/1/relationships/speakers",
                    "related": "/v1/users/1/speakers"
                }
            },
            "marketer-events": {
                "links": {
                    "self": "/v1/users/1/relationships/marketer-events",
                    "related": "/v1/events"
                }
            },
            "organizer-events": {
                "links": {
                    "self": "/v1/users/1/relationships/organizer-events",
                    "related": "/v1/events"
                }
            },
            "coorganizer-events": {
                "links": {
                    "self": "/v1/users/1/relationships/coorganizer-events",
                    "related": "/v1/events"
                }
            },
            "moderator-events": {
                "links": {
                    "self": "/v1/users/1/relationships/moderator-events",
                    "related": "/v1/events"
                }
            },
            "orders": {
                "links": {
                    "self": "/v1/users/1/relationships/orders",
                    "related": "/v1/users/1/orders"
                }
            },
            "notifications": {
                "links": {
                    "self": "/v1/users/1/relationships/notifications",
                    "related": "/v1/users/1/notifications"
                }
            }
        },
        "id": "1",
        "links": {
            "self": "/v1/users/1"
        }
    },
    "links": {
        "self": "/v1/users/1"
    },
    "jsonapi": {
        "version": "1.0"
    }
}
```

**Comment by iamareebjamal** [comment]
Yeah, all images are null. It is not even working

**Comment by Haider8** [comment]
@iamareebjamal  @CosmicCoder96 This is the JSON of the user now
```
{
    "data": {
        "type": "user",
        "attributes": {
            "original-image-url": "http://127.0.0.1:5000/static/media/events/1/original/RzZOSVN1bm/38e657d6-1758-4fc6-b5db-4ec0004c9d73.jpg",
            "facebook-id": null,
            "is-user-coorganizer": false,
            "icon-image-url": "http://127.0.0.1:5000/static/media/events/1/icon/YmZKaGZjSH/62a421d8-6843-49c6-bcce-530d99bbe7b8.jpg",
            "is-user-attendee": false,
            "last-accessed-at": "2019-03-25T04:13:17.536296+00:00",
            "contact": "7678463676",
            "instagram-url": null,
            "twitter-url": null,
            "first-name": "Haider ",
            "is-user-organizer": true,
            "avatar-url": "http://127.0.0.1:5000/static/media/temp/images/69d56829-22d4-41c5-a39d-7cd1c9676d7c/OGFkZGZkNG/99adb3f9-f1fa-4d36-9e3c-7dab0b3ea579.png",
            "deleted-at": null,
            "is-sales-admin": true,
            "was-registered-with-order": false,
            "last-name": "Ali",
            "is-user-track-organizer": false,
            "details": "User details",
            "google-plus-url": null,
            "thumbnail-image-url": "http://127.0.0.1:5000/static/media/events/1/thumbnail/TFFoR0NQb2/94586341-4070-438c-a961-6d00cdfc7e2b.jpg",
            "is-admin": true,
            "created-at": "2019-01-17T17:58:59.432218+00:00",
            "email": "haider.lee23@gmail.com",
            "is-marketer": false,
            "is-user-moderator": false,
            "facebook-url": null,
            "is-super-admin": true,
            "small-image-url": "http://127.0.0.1:5000/static/media/events/1/thumbnail/TFFoR0NQb2/94586341-4070-438c-a961-6d00cdfc7e2b.jpg",
            "is-verified": true,
            "is-user-registrar": false
        },
        "relationships": {
            "email-notifications": {
                "links": {
                    "self": "/v1/users/1/relationships/email-notifications",
                    "related": "/v1/users/1/email-notifications"
                }
            },
            "discount-codes": {
                "links": {
                    "self": "/v1/users/1/relationships/discount-codes",
                    "related": "/v1/users/1/discount-codes"
                }
            },
            "registrar-events": {
                "links": {
                    "self": "/v1/users/1/relationships/registrar-events",
                    "related": "/v1/events"
                }
            },
            "attendees": {
                "links": {
                    "self": "/v1/users/1/relationships/attendees",
                    "related": "/v1/users/1/attendees"
                }
            },
            "notifications": {
                "links": {
                    "self": "/v1/users/1/relationships/notifications",
                    "related": "/v1/users/1/notifications"
                }
            },
            "favourite-events": {
                "links": {
                    "self": "/v1/users/1/relationships/favourite-events",
                    "related": "/v1/users/1/user-favourite-events"
                }
            },
            "moderator-events": {
                "links": {
                    "self": "/v1/users/1/relationships/moderator-events",
                    "related": "/v1/events"
                }
            },
            "track-organizer-events": {
                "links": {
                    "self": "/v1/users/1/relationships/track-organizer-events",
                    "related": "/v1/events"
                }
            },
            "coorganizer-events": {
                "links": {
                    "self": "/v1/users/1/relationships/coorganizer-events",
                    "related": "/v1/events"
                }
            },
            "sales-admin-events": {
                "links": {
                    "self": "/v1/users/1/relationships/sales-admin-events",
                    "related": "/v1/events"
                }
            },
            "organizer-events": {
                "links": {
                    "self": "/v1/users/1/relationships/organizer-events",
                    "related": "/v1/events"
                }
            },
            "alternate-emails": {
                "links": {
                    "self": "/v1/users/1/relationships/alternate-emails",
                    "related": "/v1/users/1/alternate-emails"
                }
            },
            "access-codes": {
                "links": {
                    "self": "/v1/users/1/relationships/access-codes",
                    "related": "/v1/users/1/access-codes"
                }
            },
            "orders": {
                "links": {
                    "self": "/v1/users/1/relationships/orders",
                    "related": "/v1/users/1/orders"
                }
            },
            "speakers": {
                "links": {
                    "self": "/v1/users/1/relationships/speakers",
                    "related": "/v1/users/1/speakers"
                }
            },
            "events": {
                "links": {
                    "self": "/v1/users/1/relationships/events",
                    "related": "/v1/users/1/events"
                }
            },
            "sessions": {
                "links": {
                    "self": "/v1/users/1/relationships/sessions",
                    "related": "/v1/users/1/sessions"
                }
            },
            "marketer-events": {
                "links": {
                    "self": "/v1/users/1/relationships/marketer-events",
                    "related": "/v1/events"
                }
            }
        },
        "id": "1",
        "links": {
            "self": "/v1/users/1"
        }
    },
    "links": {
        "self": "/v1/users/1"
    },
    "jsonapi": {
        "version": "1.0"
    }
}
```

File: app/api/helpers/tasks.py [hunk 3]
```diff
@@ -97,6 +97,22 @@ def resize_event_images_task(self, event_id, original_image_url):
         logging.exception('Error encountered while generating resized images for event with id: {}'.format(event_id))
 
 
+@celery.task(base=RequestContextTask, name='resize.user.images', bind=True)
+def resize_user_images_task(self, user_id, original_image_url):
+    user = safe_query(db, User, 'id', user_id, 'user_id')
+    try:
+        logging.info('User image resizing tasks started {}'.format(avatar_image_url))
+        uploaded_images = create_save_image_sizes(original_image_url, 'speaker-image', user.id)
+        user.original_image_url = uploaded_images['original_image_url']
```
Comment (iamareebjamal) [inline app/api/helpers/tasks.py]: `user.avatar_url`

**Comment by Haider8** [comment]
> You are storing them in the event storage path where event images are stored. I pointed this out in my previous review. You need to store them in the user’s path.

@CosmicCoder96 @iamareebjamal Please review. Image storage path is fixed (now stores in ```user```'s path)

File: app/api/users.py [hunk 4]
```diff
@@ -87,6 +87,9 @@ def after_create_object(self, user, data, view_kwargs):
         #     del uploaded_images['large_image_url']
         #     self.session.query(User).filter_by(id=user.id).update(uploaded_images)
 
+        if data.get('original_image_url'):
```
Comment (iamareebjamal) [inline app/api/users.py]: avatar_url

**Comment by iamareebjamal** [comment]
Also, add the JSON response of the user

**Comment by Haider8** [comment]
> Also, add the JSON response of the user

@iamareebjamal 
```
{
    "data": {
        "type": "user",
        "relationships": {
            "alternate-emails": {
                "links": {
                    "self": "/v1/users/3/relationships/alternate-emails",
                    "related": "/v1/users/3/alternate-emails"
                }
            },
            "organizer-events": {
                "links": {
                    "self": "/v1/users/3/relationships/organizer-events",
                    "related": "/v1/events"
                }
            },
            "track-organizer-events": {
                "links": {
                    "self": "/v1/users/3/relationships/track-organizer-events",
                    "related": "/v1/events"
                }
            },
            "marketer-events": {
                "links": {
                    "self": "/v1/users/3/relationships/marketer-events",
                    "related": "/v1/events"
                }
            },
            "orders": {
                "links": {
                    "self": "/v1/users/3/relationships/orders",
                    "related": "/v1/users/3/orders"
                }
            },
            "discount-codes": {
                "links": {
                    "self": "/v1/users/3/relationships/discount-codes",
                    "related": "/v1/users/3/discount-codes"
                }
            },
            "registrar-events": {
                "links": {
                    "self": "/v1/users/3/relationships/registrar-events",
                    "related": "/v1/events"
                }
            },
            "moderator-events": {
                "links": {
                    "self": "/v1/users/3/relationships/moderator-events",
                    "related": "/v1/events"
                }
            },
            "speakers": {
                "links": {
                    "self": "/v1/users/3/relationships/speakers",
                    "related": "/v1/users/3/speakers"
                }
            },
            "events": {
                "links": {
                    "self": "/v1/users/3/relationships/events",
                    "related": "/v1/users/3/events"
                }
            },
            "favourite-events": {
                "links": {
                    "self": "/v1/users/3/relationships/favourite-events",
                    "related": "/v1/users/3/user-favourite-events"
                }
            },
            "access-codes": {
                "links": {
                    "self": "/v1/users/3/relationships/access-codes",
                    "related": "/v1/users/3/access-codes"
                }
            },
            "attendees": {
                "links": {
                    "self": "/v1/users/3/relationships/attendees",
                    "related": "/v1/users/3/attendees"
                }
            },
            "sales-admin-events": {
                "links": {
                    "self": "/v1/users/3/relationships/sales-admin-events",
                    "related": "/v1/events"
                }
            },
            "sessions": {
                "links": {
                    "self": "/v1/users/3/relationships/sessions",
                    "related": "/v1/users/3/sessions"
                }
            },
            "email-notifications": {
                "links": {
                    "self": "/v1/users/3/relationships/email-notifications",
                    "related": "/v1/users/3/email-notifications"
                }
            },
            "coorganizer-events": {
                "links": {
                    "self": "/v1/users/3/relationships/coorganizer-events",
                    "related": "/v1/events"
                }
            },
            "notifications": {
                "links": {
                    "self": "/v1/users/3/relationships/notifications",
                    "related": "/v1/users/3/notifications"
                }
            }
        },
        "attributes": {
            "details": "",
            "small-image-url": "http://127.0.0.1:5000/static/media/users/3/thumbnail/NGRqaXMwaF/8e2be2e8-519b-4a93-869a-fe20714c10dd.jpg",
            "is-user-coorganizer": false,
            "last-accessed-at": "2019-03-26T09:41:32.620906+00:00",
            "is-admin": false,
            "first-name": "Haider",
            "is-user-registrar": false,
            "twitter-url": null,
            "contact": "67676767765435435435",
            "is-verified": false,
            "email": "test@test.io",
            "created-at": "2019-03-03T19:00:39.612676+00:00",
            "avatar-url": "http://127.0.0.1:5000/static/media/users/3/original/a01FVSs4Ml/360db41a-99e4-48f1-acee-06f891cb8e20.jpg",
            "is-super-admin": false,
            "is-user-moderator": false,
            "is-user-attendee": false,
            "google-plus-url": null,
            "is-sales-admin": false,
            "was-registered-with-order": false,
            "instagram-url": null,
            "is-user-track-organizer": false,
            "original-image-url": "http://127.0.0.1:5000/static/media/events/3/original/TEx2YTllK2/9152b1cc-4057-4327-8ebd-6c9209671f12.jpg",
            "is-user-organizer": true,
            "thumbnail-image-url": "http://127.0.0.1:5000/static/media/users/3/thumbnail/NGRqaXMwaF/8e2be2e8-519b-4a93-869a-fe20714c10dd.jpg",
            "deleted-at": null,
            "facebook-url": null,
            "is-marketer": false,
            "facebook-id": null,
            "icon-image-url": "http://127.0.0.1:5000/static/media/users/3/icon/SzJhQkk5S3/1632a0e3-df96-40cd-9fd8-cbd20484ef65.jpg",
            "last-name": "Ali"
        },
        "id": "3",
        "links": {
            "self": "/v1/users/3"
        }
    },
    "links": {
        "self": "/v1/users/3"
    },
    "jsonapi": {
        "version": "1.0"
    }
}
```

File: app/api/users.py (hunk 4 above)
Comment (Haider8) [reply app/api/users.py]: Done

**Comment by CosmicCoder96** [comment]
@Haider8 @iamareebjamal  Original Imager URL is still getting stored in the `event` directory?

**Comment by iamareebjamal** [comment]
@CosmicCoder96 It's unchanged from the previous change AFAIK, when everything was being stored in event directory

What should be done about original-image-url, is it being used anywhere?

**Comment by Haider8** [comment]
@iamareebjamal @CosmicCoder96 Can we make ```original_image_url``` and ```avatar_url``` same for now? Both will have same path and same name.

```user.original_image_url = uploaded_images['original_image_url']```

**Comment by iamareebjamal** [comment]
OK, assign same value to both. But make original_image_url dump_only

**Comment by iamareebjamal** [comment]
LGTM. @CosmicCoder96 Please review

**Pull request merged**